File: priv/combined_metrics/samples/scope_and_assignment/declared_close_to_use/good/pipeline_processor.py
"""Data processing pipeline — GOOD: variables declared immediately before use."""

from datetime import datetime, timezone
import uuid


def process_order(order):
    min_price = 0.01
    max_items = 50

    validated = [
        item
        for item in order["items"]
        if item["quantity"] > 0
        and item["price"] >= min_price
        and len(order["items"]) <= max_items
    ]

    subtotal = sum(item["price"] * item["quantity"] for item in validated)

    discount_threshold = 100
    premium_discount = 0.15
    standard_discount = 0.05

    if subtotal > discount_threshold:
        discount = subtotal * premium_discount
    else:
        discount = subtotal * standard_discount

    discounted = subtotal - discount
    tax_rate = 0.08
    tax = discounted * tax_rate
    total = discounted + tax

    currency = "USD"
    return {"total": total, "currency": currency, "item_count": len(validated)}


def process_batch(orders):
    max_batch_size = 200

    if len(orders) > max_batch_size:
        return ("batch_error", "too_large")

    results = []
    for order in orders:
        outcome = process_order(order)
        if outcome["total"] > 0:
            results.append(("ok", outcome["total"]))
        else:
            results.append(("batch_error", order["id"]))

    successes = sum(1 for status, _ in results if status == "ok")
    batch_id = uuid.uuid4().hex
    started_at = datetime.now(timezone.utc)

    return {
        "batch_id": batch_id,
        "started_at": started_at,
        "total": len(orders),
        "successes": successes,
    }


def summarize(results):
    lines = [f"{status}: {val}" for status, val in results]
    body = "\n".join(lines)

    label = "Summary"
    separator = "-" * 40

    return f"{label}\n{separator}\n{body}"

File: priv/combined_metrics/samples/scope_and_assignment/declared_close_to_use/good/pipeline_processor.rb
# Data processing pipeline — GOOD: variables declared immediately before use.

require "securerandom"
require "time"

module Pipeline
  module Processor
    module_function

    def process_order(order)
      min_price = 0.01
      max_items = 50

      validated = order[:items].select do |item|
        item[:quantity] > 0 &&
          item[:price] >= min_price &&
          order[:items].length <= max_items
      end

      subtotal = validated.reduce(0) { |acc, item| acc + item[:price] * item[:quantity] }

      discount_threshold = 100
      premium_discount = 0.15
      standard_discount = 0.05

      discount =
        if subtotal > discount_threshold
          subtotal * premium_discount
        else
          subtotal * standard_discount
        end

      discounted = subtotal - discount
      tax_rate = 0.08
      tax = discounted * tax_rate
      total = discounted + tax

      currency = "USD"
      { total: total, currency: currency, item_count: validated.length }
    end

    def process_batch(orders)
      max_batch_size = 200

      return [:batch_error, :too_large] if orders.length > max_batch_size

      results = orders.map do |order|
        outcome = process_order(order)
        outcome[:total] > 0 ? [:ok, outcome[:total]] : [:batch_error, order[:id]]
      end

      successes = results.count { |status, _| status == :ok }
      batch_id = SecureRandom.uuid
      started_at = Time.now.utc

      {
        batch_id: batch_id,
        started_at: started_at,
        total: orders.length,
        successes: successes
      }
    end

    def summarize(results)
      lines = results.map { |status, val| "#{status}: #{val}" }
      body = lines.join("\n")

      label = "Summary"
      separator = "-" * 40

      "#{label}\n#{separator}\n#{body}"
    end
  end
end
